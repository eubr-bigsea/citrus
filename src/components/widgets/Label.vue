<template>
    <span class="label">
        <span :data-order="field.order" :data-name="field.name"
            :class="{'text-danger': error, required: field.required}" :title="error">{{field.label || field.name}}</span>
        <span v-show="field.required">*</span>:
        <span v-if="(showHelp === undefined || showHelp) && field.help" class="data-help">
            <span class="fa fa-question-circle float-right" :title="field.help"></span>
        </span>
    </span>
</template>

<script>
    import Widget from '../../mixins/Widget.js';
    export default {
        mixins: [Widget],
        name: "property-label",
        props: {
            field: { required: true, type: Object },
            value: {},
            showHelp: true,
        },
        computed: {
            error() {
                if (this.field.required && (this.value === undefined || this.value === null || this.value === '' || this.value === {})) {
                    return this.$tc('errors.missingRequiredValue');
                }
                return null;
            },
        }
    }
</script>
<style scoped>
    .label {
        font-size: 1em;
    }
</style>
