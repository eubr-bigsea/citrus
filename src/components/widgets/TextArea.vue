<template>
    <div>
        <div v-if="readOnly">
            {{value === null ? field.default: value}}
        </div>
        <div v-else>
            <LabelComponent :field="field" :value="value" :show-help="showHelp"></LabelComponent>
            <textarea class="form-control input-sm" @keyup="updated" autocomplete="off" autocorrect="off"
                autocapitalize="off" spellcheck="false" :value="value === null ? field.default: value"
                rows="3"></textarea>
        </div>
    </div>
</template>
<script>
    import LabelComponent from './Label.vue'
    import Widget from '../../mixins/Widget.js';
    export default {
        mixins: [Widget],
        components: { LabelComponent },
        methods: {
            updated: _.debounce(function (e) {
                this.triggerUpdateEvent(this.message, this.field, e.target.value);
            }, 500)
        },
        computed: {
            normalizedValue: () => {
                return this.field.value || this.field.default;
            }
        },
    }
</script>