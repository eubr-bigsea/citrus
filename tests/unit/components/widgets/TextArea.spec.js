import { shallowMount } from '@vue/test-utils';
import TextArea from '$SRC/components/widgets/TextArea.vue';
import Label from '$SRC/components/widgets/Label.vue';
import { expect } from 'vitest';

describe('TextArea.vue', () => {
    afterEach(() => {
        vi.restoreAllMocks();
        vi.useRealTimers();
    });
    beforeEach(() => {
        vi.useFakeTimers({ toFake: ['setTimeout', 'clearTimeout', 'Date'] });
    });
    it('renders read-only widget with default value', () => {
        const wrapper = shallowMount(TextArea, {
            propsData: {
                field: {
                    value: null,
                    default: 'User input'
                },
                readOnly: true,
                value: null,
            }
        });
        expect(wrapper.findComponent(Label).exists()).toBe(false);
        expect(wrapper.text()).toBe("User input");
    });
    it('renders with field value', () => {
        const wrapper = shallowMount(TextArea, {
            propsData: {
                field: {
                    value: 3001,
                    default: 0
                },
                readOnly: false,
                value: null
            }
        });
        expect(wrapper.findComponent(Label).exists()).toBeTruthy();
        const textarea = wrapper.find('textarea');
        expect(textarea.exists()).toBeTruthy();

    });
    it('renders and emits after input ', async () => {
        const spy = vi.spyOn(TextArea.methods, 'updated');
        const wrapper = shallowMount(TextArea, {
            propsData: {
                field: {
                    value: 'Text', default: 0
                },
                readOnly: false, value: null
            }
        });

        const textarea = wrapper.find('textarea');
        const text = 'New data for property';
        textarea.setValue(text);
        textarea.trigger('keyup');
        vi.advanceTimersByTime(2000);
        const payload = wrapper.emitted();
        expect(payload.update[0][1]).toBe(text);
        expect(spy).toBeCalled();

    });
    it('snapshots the component', () => {
        const wrapper = shallowMount(TextArea, { propsData: { field: {} } });
        expect(wrapper.text()).toMatchSnapshot();
    });
});
