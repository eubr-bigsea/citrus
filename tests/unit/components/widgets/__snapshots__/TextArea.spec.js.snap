// Vitest Snapshot v1

exports[`Text.vue > snapshots the component 1`] = `""`;

exports[`TextArea.vue > snapshots the component 1`] = `""`;
